packages/components/src/FormField/FormFieldWrapper.tsx

```diff
@@ -0,0 +1,125 @@
+import { InputValidation } from "../InputValidation";
+
+interface FormFieldWrapperProps extends FormFieldProps {
+  error: string;
+  identifier: string;
+}
+
+interface LabelPadding {
+  paddingLeft: number | string | undefined;
+  paddingRight: number | string | undefined;
+}
+
+export function FormFieldWrapper({
+  align,
+  placeholder,
+  value,
+  children,
+  invalid,
+  error,
+  size,
+  prefix,
+  suffix,
+  max,
+  maxLength,
+  type,
+  disabled,
+  inline,
+  identifier,
+}: PropsWithChildren<FormFieldWrapperProps>) {
+  const wrapperClasses = classnames(
+    styles.wrapper,
+    size && styles[size],
+    align && styles[align],
+    {
+      [styles.miniLabel]:
+        (placeholder && value !== "") || (placeholder && type === "select"),
+      [styles.textarea]: type === "textarea",
+      [styles.invalid]: invalid ?? error,
+      [styles.disabled]: disabled,
+      [styles.inline]: inline,
+      [styles.maxLength]: maxLength,
+      [styles.select]: type === "select",
+    },
+  );
+
+  const wrapperInlineStyle = {
+    ["--formField-maxLength" as string]: maxLength || max,
+  };
+
+  const prefixRef = useRef() as RefObject<HTMLDivElement>;
+  const suffixRef = useRef() as RefObject<HTMLDivElement>;
+
+  const [labelStyle, setLabelStyle] = useState<LabelPadding>({
+    paddingLeft: undefined,
+    paddingRight: undefined,
+  });
+
+  useEffect(() => {
+    setLabelStyle(getAffixPaddding);
+  }, [value]);
+
+  return (
+    <>
+      <div className={wrapperClasses} style={wrapperInlineStyle}>
+        {prefix?.icon && <AffixIcon {...prefix} size={size} />}
+        <div className={styles.inputWrapper}>
+          <label
+            className={styles.label}
+            htmlFor={identifier}
+            style={
+              prefixRef?.current || suffixRef?.current ? labelStyle : undefined
+            }
+          >
+            {placeholder}
+          </label>
+          {prefix?.label && <AffixLabel {...prefix} labelRef={prefixRef} />}
+          {children}
+          {suffix?.label && (
+            <AffixLabel {...suffix} labelRef={suffixRef} variation="suffix" />
+          )}
+        </div>
+        {suffix?.icon && (
+          <AffixIcon {...suffix} variation="suffix" size={size} />
+        )}
+      </div>
+      {error && !inline && <InputValidation message={error} />}
+    </>
+  );
+
+  function getAffixPaddding() {
+    const newPadding: LabelPadding = {
+      paddingLeft: undefined,
+      paddingRight: undefined,
+    };
+
+    if (prefixRef?.current) {
+      const { offsetWidth } = prefixRef?.current;
+      newPadding.paddingLeft =
+        value !== ""
+          ? undefined
+          : `calc(${offsetWidth}px + var(--space-small))`;
+    }
+
+    if (suffixRef?.current) {
+      const { offsetWidth } = suffixRef?.current;
+      newPadding.paddingRight =
+        value !== ""
+          ? undefined
+          : `calc(${offsetWidth}px + var(--space-small))`;
+    }
+
+    return newPadding;
+  }
```

Comment:
I will try my best to explain. If i fail let me know and we can chat about this.

---

Because the `placeholder` needs to sit at the beginning of the `input` when there is no value, and move above the `prefix` when there is a value, we found some challenges. 

To animate the `placeholder` was tough because we never knew how long the prefix would be, so the `left` position of the `placeholder` would need to be different if the prefix was `$` or the prefix was `Search`. The solution is to find the element on the page, measure its width and adjust accordingly. This means that the length of `prefix` doesn't really matter.

This then changed from adjusting the `left` property to `padding` to help for text alignment.

Reply:
👍  explanation makes sense!